# Description: 
This data is about housing includes the information about the size of the house, number of bedrooms, the location and a few other features. The goal is to predict the price of the house based on these features. 

Using 30/70 test/train split(30% test and 70% train):
1) Implement a gradient descent algorithm that can do regression and predicts the price. 
Report the mean squared error for train and test. 
2) Choose a threshold T and assume all the houses with a price larger than T are labeled as expensive and the rest are labeled as non-expensive. Implement a simple perceptron that can classify the houses as expensive/non-expensive for you. Report accuracy for train and test.
You are free to use any language, but the algorithms should be implemented by yourself. Data is attached.

Please make a folder in your Github repository for this assignment and commit your code there. Then post the link to the repository + results, in canvas. 



1) Implement a gradient descent algorithm that can do regression and predicts the price. Report the mean squared error for train and test. 

In [1]:
#computes output vector by finding the dot product of each row of features by the weights
def compute_output(w,x):
    output = np.dot(x,w.T)
    return output

In [2]:
#computes squarred error by summing the error and dividing by 2
def get_squared_error(error):
    error = np.square(error)
    sum_error = error.sum()
    sum_error = sum_error/np.size(error)
    return sum_error

In [3]:
#implement gradient descent
import numpy as np
def findModel(x,y):
    w = np.random.rand(5) #initial weights
    w = np.matrix(w)
    R = .001 # learning rate
    output = compute_output(w,x) #initial price predictions
    iterations = 0 #iteration count
    max_it = 10000000 #maximum number of iterations
    squared_error = 10000000 # arbitrary initial squared error
    tol = .01
    prev_errors = []
    while((squared_error > tol) and (iterations < max_it)): #gradient descent
        errors= np.array(output) - np.array(y)
        squared_error = get_squared_error(errors)#get squared error
        prev_errors.append(squared_error)
        # dE/dwi = sum(error[j]*-x[i][j]), where j is example, and i is features column
        gradient_E = x.T.dot(errors) #should be a 5x1 vector of dE/dwi
        #update weights
        w = w.T - (R* gradient_E)
        w = w.T
        iterations += 1#increase iteration count
        output = compute_output(w,x) #compute output with new weights
    return w

In [4]:
import numpy as np
from math import sqrt
#read in features
f = open("housing_prices.txt")
lines = f.readlines()
lines = lines[1:]
features = [[]]*(len(lines))
count = 0
for line in lines:
    features[count] = line.split()
    tempStreet = ""
    stringcount = 0
    for f in features[count]:
        try:
            f = float(f)
        except ValueError:
            tempStreet += f + " "
            stringcount +=1
    features[count].append(tempStreet)
    for i in range(stringcount):
        features[count].pop(2)
    count += 1
y = [] #output
x = [] #input
for f in features:
    y.append(float(f[1]))
    f.pop(1)
    f[0]= float(f[0])
    f[1]= float(f[1])
    f[2]= float(f[2])
    x.append(f)
        
#make dictionary to map street names to values
#replace street names with values
street_names = {}
count3 = 1
for i in range(len(x)):
    if x[i][3] not in street_names.keys():
        street_names[x[i][3]] = count3
        count3 += 1
    x[i][3] = street_names[x[i][3]]
    x[i] = [1] + x[i]#create x0 column
#rescale features
#using x' = (x - min)/(max-min)
x1 = []
x2 = []
x3 = []
x4 = []
for i in range(len(x)):
    x1.append(x[i][1])
    x2.append(x[i][2])
    x3.append(x[i][3])
    x4.append(x[i][4])
#store pre-normalized values
y_real = []
ymax = max(y)
ymin = min(y)
#normalize
for i in range(len(x)):
    x[i][1] = (x[i][1] - min(x1))/(max(x1)-min(x1))
    x[i][2] = (x[i][2] - min(x2))/(max(x2)-min(x2))
    x[i][3] = (x[i][3] - min(x3))/(max(x3)-min(x3))
    x[i][4] = (x[i][4] - min(x4))/(max(x4)-min(x4))
    y_real.append(y[i])
    y[i] = (y[i] - min(y))/(max(y) - min(y))
#divide train/test
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split( x, y, test_size=0.30)
x_train = np.matrix(x_train)
y_train = np.matrix(y_train)
y_train = y_train.T
x_test = np.matrix(x_test)
y_test = np.matrix(y_test)
#w = findModel(x_train,y_train)
#print(w)
w = [ 0.08793469,  0.71704488, -0.08180255,  0.00179489,  0.12905482]
w = np.matrix(w)
o = compute_output(w,x_train)
errors= o - y_train
se = get_squared_error(errors)
print(se)

0.039455445572


# For normalized data: 
since the weight vector is initialized to be random, the output of the function is different each time it's run

 w= [ 0.08793469,  0.71704488, -0.08180255,  0.00179489,  0.12905482]
 
 b = 0.08793469
 
 w1 = 0.71704488
 
 w2 = -0.08180255
 
 w3 = 0.00179489
 
 w4 = 0.12905482
 
 Train Error (MSE)= 0.039455445572

In [5]:
o_test = compute_output(w, x_test)
y_test = y_test.T
errors_test = o_test - y_test
se_test = get_squared_error(errors_test)
print(se_test)

0.0300908615666


# For normalized data:
Test Error (MSE) = 0.0300908615666


2) Choose a threshold T and assume all the houses with a price larger than T are labeled as expensive and the rest are labeled as non-expensive. Implement a simple perceptron that can classify the houses as expensive/non-expensive for you. Report accuracy for train and test. 

In [6]:
#create classifications vector
T = 500000
house_classification = []
for p in y_real:
    if p >= 500000:
        house_classification.append(1)
    else:
        house_classification.append(-1)
#train/test classifications
x_train, x_test, y_train, y_test = train_test_split( x, house_classification, test_size=0.30)
x_train = np.matrix(x_train)
y_train = np.matrix(y_train)
y_train = y_train.T
x_test = np.matrix(x_test)
y_test = np.matrix(y_test)
y_test = y_test.T

In [7]:
def classify(w,x):
    classifications = []
    output = np.dot(x,w.T)
    for o in output:
        if o >= 0:
            classifications.append(1)
        else:
            classifications.append(-1)
    return np.matrix(classifications)

In [8]:
def get_accuracy(error):
    #tp+tn/total
    total = np.size(error)
    tptn = 0
    error_array = np.array(error)
    for i in error_array:
        if i == 0:
            tptn += 1
    accuracy = tptn/total
    return accuracy

In [12]:
def findModelClassify(x,y):
    w = np.random.rand(5) #initial weights
    w = np.matrix(w)
    R = .0005 # learning rate
    output = classify(w,x) #initial price predictions
    iterations = 0 #iteration count
    max_it = 100000 #maximum number of iterations
    accuracy  = 0 # arbitrary initial squared error
    tol = 1
    prev_errors = []
    while((accuracy < tol) and (iterations < max_it)): #gradient descent
        errors= np.array(output.T) - np.array(y)
        accuracy = get_accuracy(errors)#get accuracy
        # dE/dwi = sum(error[j]*-x[i][j]), where j is example, and i is features column
        gradient_E = np.dot(x.T,errors) #should be a 5x1 vector of dE/dwi
        #update weights
        w = w.T - (R* gradient_E)
        w = w.T
        iterations += 1#increase iteration count
        output = classify(w,x) #compute output with new weights
    print(accuracy)
    return w

In [13]:
w = findModelClassify(x_train,y_train)
print(w)

0.603448275862069
[[-0.01155368  0.06713061 -0.00832529 -0.0006586  -0.00130368]]


In [15]:
o_test = classify(w,x_test)
test_errors= np.array(o_test.T) - np.array(y_test)
test_accuracy = get_accuracy(test_errors)
print(test_accuracy)

0.6


# Train/Test Error

For train: 

accuracy = 0.603448275862069

For test:

accuracy = 0.6